bundles/org.openhab.ui.dashboard/src/main/java/org/openhab/ui/dashboard/internal/DashboardService.java

```diff
@@ -220,15 +227,57 @@ private void addTilesForExternalServices(Map<String, Object> properties) {
                     String url = (String) properties.get(linkname + LINK_URL);
                     String imageUrl = (String) properties.get(linkname + LINK_IMAGEURL);
 
-                    logger.debug("Add link: {}", name);
+                    url = fillMagicWords(url);
+                    imageUrl = fillMagicWords(imageUrl);
+
+                    if (url != null && !url.isEmpty()) {
+                        logger.debug("Add link: {}", name);
 
-                    addDashboardTile(new ExternalServiceTile.DashboardTileBuilder().withName(name).withUrl(url)
-                            .withImageUrl(imageUrl).build());
+                        addDashboardTile(new ExternalServiceTile.DashboardTileBuilder().withName(name).withUrl(url)
+                                .withImageUrl(imageUrl).build());
+                    } else {
+                        logger.warn("Ignoring empty URL link for tile '{}'", name);
+                    }
                 }
             }
         }
     }
 
+    private @Nullable String fillMagicWords(@Nullable String linkname) {
```

Comment:
Should we maybe inspect headers such as the x forwarded for, see for example the nginx example https://www.nginx.com/resources/wiki/start/topics/examples/likeapache/

Reply:
Good idea @martinvw, but those are just pure url links. Dashboard does not forward anything, it just  create tile which have static url link. Or did I miss you point?

Reply:
If running behind a reverse proxy [as described here](https://docs.openhab.org/installation/security.html#nginx-reverse-proxy) the server host name is not relevant its about the hostname on which its lives according to users reaching it from outside.

Lets say I run my openhab at home on my server at elbrus.mtin.localhost:5800 externally I want to expose it more friendly something like openhab.martinvw.nl maybe ;-)

If I reach the page remotely I would like it to replace those links with openhab.martinvw.nl and not with elbrus.mtin.localhost:5800. 

BTW, would this problem even exist with relative links?

Reply:
I understand proxy use case, but I don't understand how this could be solved with dashboard url's. Dashboard static tiles are generated during OH startup, so there is no mechanism to dynamically change the url. If OH is accessed externally, tile url should always use full domain name (e.g. openhab.martinvw.nl) or user should add two tiles, one for local access and one for external access.

>BTW, would this problem even exist with relative links?

I'm not fully sure, but I have feeling that relative links don't work with different tcp ports.

Reply:
You are right about the ports.

@ThomDietrich could you elaborate a little more on the use case so that we know whether the solution of Pali suffices or that we need to take reverse proxies into account as well.

Reply:
Imho there are two degrees of how my initial request can be implemented

1. the option to have a placeholder for the hostname to solve the issues openHABian users were seeing in their local subnet after changing the hostname of their system.

2. the option to have a placeholder which represents the hostname/domain name through which the user is currently accessing the dashboard, e.g. for users with a reverse proxy setup.

I totally agree that option 2 would be a great addition! From our short discussion in #272 I got the idea that @paulianttila was not sure how to implement that request. The comment above supports that.

@paulianttila just a quick idea. To work around the issue of having the tile URLs build statically during startup, how about an active javascript solution during runtime? As you probably know, the options are exhaustive: https://stackoverflow.com/a/17336519

Reply:
@martinvw, do you mean that when user's web browser (client) access dashboard which is behind proxy, proxy add X-Forwarded headers to request. Dashboard bundle recognise that http request contains X-Forwarded headers and provide different tile URLs than without X-Forwarded headers?

Reply:
Correct that is imho how it should be handled.

Reply:
@martinvw, that’s imho proxy feature and in fact should be possible already e.g. with nginx.

Reply:
But does the dashboard also do what you describe?

> Dashboard bundle recognise that http request contains X-Forwarded headers and provide different tile URLs than without X-Forwarded headers?

Reply:
No, that was my understanding how it should work from you point of view :)

Reply:
That might work out indeed.

I must say that I also never saw it in action, so I will aim on installing it tonight, but should I be able to install it through the paperUI? I cannot see it there?

Reply:
That client side javascript doesn't solve the problem, imho it just split current page url to different parts. Or is there some kind of trick on you mind which I didn't understand?

I try to summaries the problem what this PR tries to solve. User install openhab and e.g. frontail via openhabian. Openhabian add frontail link to dashboard. Currently openhabian create link during install which contains current host name (e.g. openhab.local) to dashboard link. If user change the host name later (e.g. to openhab2.local), dashboard link doesn't work anymore as link refers to old host name (openhab.local rather than openhab2.local). This PR make possible to openhabian to write $HOSTNAME special word to dashboard link, which dashboard bundle will replace by the current host name. So links should work even user change the host name later.

This reverse proxy thing is a different problem and imho should be solved via nginx. See http://nginx.org/en/docs/http/ngx_http_sub_module.html. By that feature nginx could replace dashboard links from local network name to remote name (@martinvw's example elbrus.mtin.localhost:5800 -> openhab.martinvw.nl).

Reply:
Thanks for the summary I really missed these details, but I'm personally a little bit in doubt whether this is the responsibility of the dashboard binding. But if you feel that this is generic and not specific for one goal I think that I agree.

Reply:
@paulianttila your analysis is correct. The problem I am seeing is, that your current implementation will replace the dashboard URL placeholder with *some* fixed address that might be wrong in *some* moment in time (e.g. Martin might want to access his setup locally AND via the public domain).

That is why I propose to not replace the placeholder the way you are currently doing it but instead add a client-side javascript to do so. That means, that a user's browser will first receive the URL with the placeholder still in place, then the javascript will generate the correct URL matching the current `window.location.hostname` and replace the href of the dashboard tile. The link above didn't show a complete example of how to achieve that, the first answer here looks like 99% percent of the solution: https://stackoverflow.com/questions/7449480/change-domain-portion-of-links-with-javascript-or-jquery

Reply:
@ThomDietrich, now I got you idea :)

I committed a prototype. I'm not a jquery guru, so most probably there is more elegant way...

Console output:
url: localhost:8080
orginal onclick: location.href = 'http://{HOSTNAME}:1234'
modified onclick: location.href = 'http://localhost:1234'
orginal onclick: location.href = 'http://{HOST}'
modified onclick: location.href = 'http://localhost:8080'

Reply:
> I'm sorry, what is "it"?

The following:

> Openhabian add frontail link to dashboard. Currently openhabian create link during install which contains current host name (e.g. openhab.local) to dashboard link. If user change the host name later (e.g. to openhab2.local), dashboard link doesn't work anymore as link refers to old host name (openhab.local rather than openhab2.local).

So a fixed link is added by openhabian, hostname is changed and then it does not work anymore.